File: examples/ios/MentraExample/ConsoleScreen.swift
import SwiftUI

struct ConsoleScreen: View {
    @ObservedObject var model: BluetoothViewModel
    @Environment(\.keyboardVisible) private var keyboardVisible
    @State private var filter = "ALL"

    private var filteredEvents: [ExampleEvent] {
        filter == "ALL" ? model.events : model.events.filter { $0.tag == filter }
    }

    var body: some View {
        ScrollView {
            VStack(spacing: 0) {
                PageHeader(title: "Console", connected: model.glassesConnected)

                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 6) {
                        Button { filter = "ALL" } label: {
                            HStack(spacing: 6) {
                                Text("ALL").font(.system(size: 11, weight: .bold)).tracking(0.5).foregroundColor(.white)
                                Text("\(model.events.count)").font(.system(size: 10, weight: .medium)).foregroundColor(Color.white.opacity(0.5))
                            }
                            .padding(.horizontal, 12).padding(.vertical, 7)
                            .background(LinearGradient(colors: [Color(hex: 0x28473A), Color(hex: 0x1F3A2A)], startPoint: .top, endPoint: .bottom))
                            .clipShape(Capsule())
                        }
                        FilterChip(color: Color(hex: 0x00C7BE), labelColor: Color(hex: 0x00807B), label: "LIVE", count: "\(model.events.filter { $0.tag == "LIVE" }.count)", active: filter == "LIVE") { filter = "LIVE" }
                        FilterChip(color: Color(hex: 0x84B5E8), labelColor: Color(hex: 0x3478B8), label: "BLE", count: "\(model.events.filter { $0.tag == "BLE" }.count)", active: filter == "BLE") { filter = "BLE" }
                        FilterChip(color: AppColor.amber, labelColor: Color(hex: 0xB86A00), label: "TX", count: "\(model.events.filter { $0.tag == "TX" }.count)", active: filter == "TX") { filter = "TX" }
                        FilterChip(color: AppColor.gold, labelColor: Color(hex: 0x8C7400), label: "STORE", count: "\(model.events.filter { $0.tag == "STORE" }.count)", active: filter == "STORE") { filter = "STORE" }
                    }
                    .padding(.horizontal, 16)
                }
                .padding(.top, 8)

                consoleCard.padding(.horizontal, 16).padding(.top, 12)
                jsonCard.padding(.horizontal, 16).padding(.top, 12)
            }
            .padding(.bottom, LayoutMetric.scrollBottomPadding(keyboardVisible: keyboardVisible))
        }
        .background(AppColor.bg)
        .scrollDismissesKeyboard(.interactively)
    }

    private var consoleCard: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                HStack(spacing: 8) {
                    HStack(spacing: 4) {
                        Circle().fill(Color(hex: 0xFF5F57)).frame(width: 8, height: 8)
                        Circle().fill(Color(hex: 0xFEBC2E)).frame(width: 8, height: 8)
                        Circle().fill(Color(hex: 0x27C93F)).frame(width: 8, height: 8)
                    }
                    Text("mentra-sdk · live").font(.system(size: 11, weight: .medium)).foregroundColor(Color.white.opacity(0.5))
                }
                Spacer()
                HStack(spacing: 5) {
                    Circle().fill(AppColor.greenSoft).frame(width: 5, height: 5)
                    Text("REC").font(.system(size: 9, weight: .bold)).tracking(0.7).foregroundColor(AppColor.greenSoft)
                }
                .padding(.horizontal, 9).padding(.vertical, 4)
                .background(AppColor.greenSoft.opacity(0.14))
                .overlay(Capsule().stroke(AppColor.greenSoft.opacity(0.3), lineWidth: 1))
                .clipShape(Capsule())
            }
            .padding(.bottom, 6)
            .overlay(alignment: .bottom) { Rectangle().fill(Color.white.opacity(0.06)).frame(height: 1) }

            VStack(spacing: 10) {
                ForEach(filteredEvents) { e in
                    HStack(alignment: .top, spacing: 10) {
                        Text(e.time)
                            .font(.system(size: 10, design: .monospaced))
                            .foregroundColor(Color.white.opacity(0.4))
                            .frame(width: 50, alignment: .leading)
                            .padding(.top, 2)
                        Text(e.tag)
                            .font(.system(size: 9, weight: .bold)).tracking(0.5)
                            .foregroundColor(tagColor(e.tag))
                            .frame(width: 50)
                            .padding(.vertical, 3)
                            .background(tagColor(e.tag).opacity(0.16))
                            .clipShape(RoundedRectangle(cornerRadius: 5))
                        Text(e.text)
                            .font(.system(size: 11, design: .monospaced))
                            .foregroundColor(AppColor.consoleText)
                        Spacer()
                    }
                }
            }
        }
        .padding(18)
        .background(AppColor.consoleBg)
        .overlay(RoundedRectangle(cornerRadius: 24).stroke(Color.white.opacity(0.06), lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 24))
        .shadow(color: Color(hex: 0x0F2A1D).opacity(0.18), radius: 40, x: 0, y: 12)
    }

    private var jsonCard: some View {
        VStack(spacing: 10) {
            Button {
                model.rawJsonExpanded.toggle()
            } label: {
                HStack(spacing: 12) {
                    Image(systemName: "chevron.left.forwardslash.chevron.right")
                        .font(.system(size: 12, weight: .semibold)).foregroundColor(AppColor.muted)
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Typed SDK status").font(.system(size: 13, weight: .semibold)).foregroundColor(AppColor.ink)
                        Text("glassesStatus, bluetoothStatus").font(.system(size: 10)).foregroundColor(AppColor.muted)
                    }
                    Spacer()
                    Image(systemName: "chevron.down").font(.system(size: 12, weight: .heavy)).foregroundColor(AppColor.ink)
                }
                .padding(.vertical, 14).padding(.horizontal, 16)
                .background(LinearGradient(colors: [Color.white.opacity(0.7), Color.white.opacity(0.5)], startPoint: .top, endPoint: .bottom))
                .overlay(RoundedRectangle(cornerRadius: 18).stroke(Color.white.opacity(0.7), lineWidth: 1))
                .clipShape(RoundedRectangle(cornerRadius: 18))
            }
            if model.rawJsonExpanded {
                Text("glassesStatus=\(String(describing: model.glassesValues))\nbluetoothStatus=\(String(describing: model.bluetoothValues))")
                    .font(.system(size: 10, design: .monospaced))
                    .foregroundColor(AppColor.consoleText)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(12)
                    .background(AppColor.consoleBg)
                    .clipShape(RoundedRectangle(cornerRadius: 14))
            }
        }
    }
}

private func tagColor(_ tag: String) -> Color {
    switch tag {
    case "BLE": Color(hex: 0x84B5E8)
    case "STORE": Color(hex: 0xE8C66B)
    case "TX": Color(hex: 0xE89C7D)
    default: Color(hex: 0x7DD89E)
    }
}

struct FilterChip: View {
    let color: Color; let labelColor: Color; let label: String; let count: String
    var active = false
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 6) {
                Circle().fill(color).frame(width: 6, height: 6)
                Text(label).font(.system(size: 11, weight: .semibold)).tracking(0.5).foregroundColor(labelColor)
                Text(count).font(.system(size: 10, weight: .medium)).foregroundColor(AppColor.muted)
            }
            .padding(.horizontal, 12).padding(.vertical, 7)
            .background(active ? color.opacity(0.14) : Color.white.opacity(0.6))
            .overlay(Capsule().stroke(active ? color.opacity(0.35) : Color.white.opacity(0.75), lineWidth: 1))
            .clipShape(Capsule())
        }
        .buttonStyle(.plain)
    }
}

File: examples/android/app/src/main/java/com/mentra/examples/android/screens/ConsoleScreen.kt
package com.mentra.examples.android.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Code
import androidx.compose.material.icons.outlined.KeyboardArrowDown
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.mentra.examples.android.ExampleEvent
import com.mentra.examples.android.MentraExampleController
import com.mentra.examples.android.isGlassesConnected
import com.mentra.examples.android.ui.AppColor
import com.mentra.examples.android.ui.PageHeader
import com.mentra.examples.android.ui.scrollBottomPadding

@Composable
fun ConsoleScreen(controller: MentraExampleController) {
    val state = controller.state
    var filter by remember { mutableStateOf("ALL") }
    val allEvents = state.events
    val events = if (filter == "ALL") allEvents else allEvents.filter { it.tag == filter }
    Column(modifier = Modifier.fillMaxSize().background(AppColor.bg).verticalScroll(rememberScrollState())) {
        PageHeader("Console")

        // Filter chips
        Row(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 4.dp).horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(6.dp)
        ) {
            Row(
                modifier = Modifier.clip(RoundedCornerShape(999.dp))
                    .background(Brush.verticalGradient(listOf(Color(0xFF28473A), Color(0xFF1F3A2A))))
                    .clickable { filter = "ALL" }
                    .padding(horizontal = 12.dp, vertical = 7.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp)
            ) {
                Text("ALL", color = Color.White, fontSize = 11.sp, fontWeight = FontWeight.Bold, letterSpacing = 0.5.sp)
                Text(allEvents.size.toString(), color = Color.White.copy(alpha = 0.5f), fontSize = 10.sp, fontWeight = FontWeight.Medium)
            }
            FilterChip(Color(0xFF00C7BE), Color(0xFF00807B), "LIVE", allEvents.count { it.tag == "LIVE" }.toString(), filter == "LIVE") { filter = "LIVE" }
            FilterChip(Color(0xFF84B5E8), Color(0xFF3478B8), "BLE", allEvents.count { it.tag == "BLE" }.toString(), filter == "BLE") { filter = "BLE" }
            FilterChip(AppColor.amber, Color(0xFFB86A00), "TX", allEvents.count { it.tag == "TX" }.toString(), filter == "TX") { filter = "TX" }
            FilterChip(AppColor.gold, Color(0xFF8C7400), "STORE", allEvents.count { it.tag == "STORE" }.toString(), filter == "STORE") { filter = "STORE" }
        }

        // Console card
        Column(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 12.dp)
                .clip(RoundedCornerShape(24.dp))
                .background(AppColor.consoleBg)
                .border(1.dp, Color.White.copy(alpha = 0.06f), RoundedCornerShape(24.dp))
                .padding(18.dp)
        ) {
            Row(modifier = Modifier.fillMaxWidth().padding(bottom = 6.dp), horizontalArrangement = Arrangement.SpaceBetween, verticalAlignment = Alignment.CenterVertically) {
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                        Box(modifier = Modifier.size(8.dp).clip(CircleShape).background(Color(0xFFFF5F57)))
                        Box(modifier = Modifier.size(8.dp).clip(CircleShape).background(Color(0xFFFEBC2E)))
                        Box(modifier = Modifier.size(8.dp).clip(CircleShape).background(Color(0xFF27C93F)))
                    }
                    Text("mentra-sdk · live", color = Color.White.copy(alpha = 0.5f), fontSize = 11.sp, fontWeight = FontWeight.Medium)
                }
                Row(
                    modifier = Modifier.clip(RoundedCornerShape(999.dp)).background(AppColor.greenSoft.copy(alpha = 0.14f))
                        .border(1.dp, AppColor.greenSoft.copy(alpha = 0.3f), RoundedCornerShape(999.dp))
                        .padding(horizontal = 9.dp, vertical = 4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(5.dp)
                ) {
                    Box(modifier = Modifier.size(5.dp).clip(CircleShape).background(AppColor.greenSoft))
                    Text("REC", color = AppColor.greenSoft, fontSize = 9.sp, fontWeight = FontWeight.Bold, letterSpacing = 0.7.sp)
                }
            }
            Box(modifier = Modifier.fillMaxWidth().height(1.dp).background(Color.White.copy(alpha = 0.06f)))
            Spacer(Modifier.height(12.dp))
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                events.forEach { e ->
                    val tagColor = tagColor(e)
                    Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                        Text(
                            e.time,
                            color = Color.White.copy(alpha = 0.4f),
                            fontSize = 10.sp,
                            fontWeight = FontWeight.Medium,
                            fontFamily = FontFamily.Monospace,
                            modifier = Modifier.width(50.dp).padding(top = 2.dp)
                        )
                        Box(
                            modifier = Modifier.width(50.dp).clip(RoundedCornerShape(5.dp)).background(tagColor.copy(alpha = 0.16f)).padding(vertical = 3.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            Text(e.tag, color = tagColor, fontSize = 9.sp, fontWeight = FontWeight.Bold, letterSpacing = 0.5.sp)
                        }
                        Text(e.text, color = AppColor.consoleText, fontSize = 11.sp, fontFamily = FontFamily.Monospace, modifier = Modifier.weight(1f))
                    }
                }
            }
        }

        // Raw JSON disclosure
        Row(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 4.dp).fillMaxWidth()
                .clip(RoundedCornerShape(18.dp))
                .background(Color.White.copy(alpha = 0.6f))
                .border(1.dp, AppColor.borderSoft, RoundedCornerShape(18.dp))
                .clickable { controller.toggleRawJson() }
                .padding(horizontal = 16.dp, vertical = 14.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Icon(Icons.Outlined.Code, null, tint = AppColor.muted, modifier = Modifier.size(14.dp))
            Column(modifier = Modifier.weight(1f)) {
                Text("Typed SDK status", color = AppColor.ink, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
                Text("glassesStatus, bluetoothStatus", color = AppColor.muted, fontSize = 10.sp)
            }
            Icon(Icons.Outlined.KeyboardArrowDown, null, tint = AppColor.ink, modifier = Modifier.size(14.dp))
        }

        if (state.rawJsonExpanded) {
            Text(
                "glassesStatus=${state.glassesStatus}\nbluetoothStatus=${state.bluetoothStatus}",
                color = AppColor.consoleText,
                fontSize = 10.sp,
                fontFamily = FontFamily.Monospace,
                modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
                    .clip(RoundedCornerShape(14.dp))
                    .background(AppColor.consoleBg)
                    .padding(12.dp)
            )
        }

        Spacer(Modifier.height(scrollBottomPadding()))
    }
}

private fun tagColor(event: ExampleEvent): Color =
    when (event.tag) {
        "BLE" -> Color(0xFF84B5E8)
        "STORE" -> Color(0xFFE8C66B)
        "TX" -> Color(0xFFE89C7D)
        else -> Color(0xFF7DD89E)
    }

@Composable
private fun FilterChip(dot: Color, labelColor: Color, label: String, count: String, active: Boolean, onClick: () -> Unit) {
    Row(
        modifier = Modifier.clip(RoundedCornerShape(999.dp))
            .background(if (active) dot.copy(alpha = 0.14f) else Color.White.copy(alpha = 0.6f))
            .border(1.dp, if (active) dot.copy(alpha = 0.35f) else AppColor.border, RoundedCornerShape(999.dp))
            .clickable { onClick() }
            .padding(horizontal = 12.dp, vertical = 7.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        Box(modifier = Modifier.size(6.dp).clip(CircleShape).background(dot))
        Text(label, color = labelColor, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, letterSpacing = 0.5.sp)
        Text(count, color = AppColor.muted, fontSize = 10.sp, fontWeight = FontWeight.Medium)
    }
}
